lib/health/health.go

```diff
@@ -122,32 +132,59 @@ func (aggregator *HealthAggregator) Summary() *HealthReport {
 	aggregator.mutex.Lock()
 	defer aggregator.mutex.Unlock()
 
+	var failedLivenessChecks []string
+	var failedReadinessChecks []string
+
 	// In the absence of any reporters, default to indicating that we are both live and ready.
 	summary := &HealthReport{Live: true, Ready: true}
 
 	// Now for each reporter...
 	for name, reporter := range aggregator.reporters {
-		log.WithFields(log.Fields{
-			"name":           name,
-			"reporter-state": reporter,
-		}).Debug("Detailed health state")
-
 		// Reset Live to false if that reporter is registered to report liveness and hasn't
 		// recently said that it is live.
-		if summary.Live && reporter.reports.Live && (!reporter.latest.Live ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		stillLive := reporter.latest.Live && !reporter.TimedOut()
+		if summary.Live && reporter.reports.Live && !stillLive {
 			summary.Live = false
 		}
 
 		// Reset Ready to false if that reporter is registered to report readiness and
 		// hasn't recently said that it is ready.
-		if summary.Ready && reporter.reports.Ready && (!reporter.latest.Ready ||
-			(reporter.timeout != 0 && time.Since(reporter.timestamp) > reporter.timeout)) {
+		stillReady := reporter.latest.Ready && !reporter.TimedOut()
+		if summary.Ready && reporter.reports.Ready && !stillReady {
 			summary.Ready = false
 		}
+
+		switch {
+		case reporter.reports.Live && !stillLive:
+			failedLivenessChecks = append(failedLivenessChecks, name)
+		case reporter.reports.Ready && !stillReady:
+			failedReadinessChecks = append(failedReadinessChecks, name)
+		default:
+			log.WithFields(log.Fields{
+				"name":           name,
+				"reporter-state": reporter,
+			}).Debug("Reporter is healthy")
+		}
 	}
 
-	log.WithField("summary", summary).Info("Overall health")
+	// Summary status has changed so update previous status and log.
+	if summary.Live != aggregator.lastReport.Live || summary.Ready != aggregator.lastReport.Ready {
+		aggregator.lastReport = summary
+		log.WithField("lastSummary", summary).Info("Overall health status changed")
+
+		for _, name := range failedLivenessChecks {
```

Comment:
I was thinking of logging a single log with a list of reporter names without the detailed state of each reporter.  I guess logging out each reporter gives us a lot more to go on, though so let's keep your approach of one log per reporter.  If you're going to do one log per reporter though, I'd suggest making   `failedLivenessChecks` a `[]*reporterState` so that you can just append the reporter instead of its name.  They you won't need to look it up again.